runtime/src/main/java/com/dylibso/chicory/runtime/Machine.java

```diff
@@ -1599,13 +1599,23 @@ private static void TABLE_INIT(MStack stack, Instance instance, long[] operands)
 
         var table = instance.table(tableidx);
 
-        if (size < 0
-                || elementidx > instance.elementCount()
-                || instance.element(elementidx) == null
-                || !(instance.element(elementidx) instanceof PassiveElement)
-                || elemidx + size > instance.element(elementidx).elementCount()
-                || end > table.size()) {
+        var elementCount = instance.elementCount();
+        var currentElement = instance.element(elementidx);
+        var currentElementCount =
+                (currentElement instanceof PassiveElement) ? currentElement.elementCount() : 0;
+        boolean isOutOfBounds =
+                (size < 0
+                        || elementidx > elementCount
+                        || (size > 0
+                                && (currentElement == null
+                                        || !(currentElement instanceof PassiveElement)))
+                        || elemidx + size > currentElementCount
+                        || end > table.size());
+
+        if (isOutOfBounds) {
             throw new WASMRuntimeException("out of bounds table access");
+        } else if (size == 0) {
+            return;
```

Comment:
nit: this is a little bit hard to follow and might make sense to break into some variables.